src/BreakpointViewer.cpp

```diff
@@ -30,98 +30,105 @@ BreakpointViewer::BreakpointViewer(QWidget* parent)
+	cnTableWidget->setSortingEnabled(true);
 	connect(cnTableWidget, &QTableWidget::itemChanged, this,
 	        &BreakpointViewer::changeCnTableItem);
 
-	selector[BREAKPOINT] = bpTableWidget;
-	selector[WATCHPOINT] = wpTableWidget;
-	selector[CONDITION]  = cnTableWidget;
-
-	userMode = true;
+	selector[BreakpointData::BREAKPOINT] = bpTableWidget;
+	selector[BreakpointData::WATCHPOINT] = wpTableWidget;
+	selector[BreakpointData::CONDITION]  = cnTableWidget;
 }
 
 // TODO: move the createSetCommand to a session manager
-void BreakpointViewer::createBreakpoint(Type type, int row)
+void BreakpointViewer::createBreakpoint(BreakpointData::Type type, int row)
 {
-	if (type == CONDITION) _createCondition(row);
+	if (type == BreakpointData::CONDITION) _createCondition(row);
 	else _createBreakpoint(type, row);
 }
 
-void BreakpointViewer::_createBreakpoint(Type type, int row)
+void BreakpointViewer::_createBreakpoint(BreakpointData::Type type, int row)
 {
-	assert(type != CONDITION);
-	auto table = selector[type];
-	auto model = table->model();
-	auto combo = (QComboBox*) table->indexWidget(model->index(row, WP_TYPE));
-	Breakpoint::Type type2;
-
-	if (type == WATCHPOINT) {
-		type2 = static_cast<Breakpoint::Type>(combo->currentIndex() + 1);
-	} else {
-		type2 = Breakpoint::BREAKPOINT;
-	}
+	assert(type != BreakpointData::CONDITION);
+	auto table             = selector[type];
+	auto model             = table->model();
+	auto combo             = (QComboBox*) table->indexWidget(model->index(row, WP_TYPE));
+	Breakpoint::Type type2 = type == BreakpointData::WATCHPOINT ? readComboBox(row) : Breakpoint::BREAKPOINT;
 
 	int address, endRange;
 	QString location = table->item(row, LOCATION)->text();
-	bool ok1 = processLocationField(-1, type, location, address, endRange,
-			combo ? combo->currentText() : "");
+	bool ok1 = parseLocationField(-1, type, location, address, endRange, combo ? combo->currentText() : "");
 	if (!ok1) {
-		auto sa = ScopedAssign(userMode, false);
-		setBreakpointStatus(type, row, false);
+		setBreakpointChecked(type, row, Qt::Unchecked);
 		return;
+	} else {
+		setBreakpointChecked(type, row, Qt::Checked);
 	}
 
 	QString condition = table->item(row, T_CONDITION)->text();
 
 	qint8 ps, ss;
-	bool ok2 = processSlotField(-1, table->item(row, SLOT)->text(), ps, ss);
+	bool ok2 = parseSlotField(-1, table->item(row, SLOT)->text(), ps, ss);
 	assert(ok2);
 
 	qint16 segment;
-	bool ok3 = processSegmentField(-1, table->item(row, SEGMENT)->text(), segment);
+	bool ok3 = parseSegmentField(-1, table->item(row, SEGMENT)->text(), segment);
 	assert(ok3);
 
-	const QString cmdStr = Breakpoints::createSetCommand(type2, address, ps, ss, segment,
-			endRange, condition);
+	const QString cmdStr = Breakpoints::createSetCommand(type2, address, ps, ss, segment, endRange, condition);
 
 	auto command = new Command(cmdStr,
-		[this] (const QString& result) {
+		[this, type, row] (const QString& result) {
+			QString id = result.trimmed();
```

Comment:
I haven't tested, maybe this is needed. But I'm surprised to see `trimmed()` here.
This is the result of a `debug set_bp ...` openMSX command, right?
I would expect this to return a string like `bp#1` without any whitespace.
Or is there a newline at the end (I also don't expect this, but I can be wrong).

Reply:
I checked again and it's not necessary. But on error messages it is because of the newline.